lib/azure/armrest/virtual_machine_service.rb

```diff
@@ -121,12 +121,187 @@ def stop(vmname, group = configuration.resource_group)
+      # virtual machine, along with the .status file. This does NOT delete
+      # copies of the disk.
+      #
+      # If the option to delete the entire storage account was selected, then
+      # it will not bother with deleting invidual files from the storage
+      # account first.
+      #
+      def delete_associated_disk(vm, options)
+        uri = Addressable::URI.parse(vm.properties.storage_profile.os_disk.vhd.uri)
+
+        # The uri looks like https://foo123.blob.core.windows.net/vhds/something123.vhd
+        storage_acct_name = uri.host.split('.').first     # storage name, e.g. 'foo123'
+        storage_acct_disk = File.basename(uri.to_s)       # disk name, e.g. 'something123.vhd'
+        storage_acct_path = File.dirname(uri.path)[1..-1] # container, e.g. 'vhds'
+
+        # Must find it this way because the resource group information isn't provided
+        sas = Azure::Armrest::StorageAccountService.new(configuration)
+
+        # Look for the storage account in the VM's resource group first. If
+        # it's not found, look through all the storage accounts.
+        begin
+          storage_acct_obj = sas.get(storage_acct_name, vm.resource_group)
+        rescue Azure::Armrest::NotFoundException
+          storage_acct_obj = sas.list_all.find { |s| s.name == storage_acct_name }
+        end
+
+        storage_acct_keys = sas.list_account_keys(storage_acct_obj.name, storage_acct_obj.resource_group)
+
+        # Deleting the storage account does not require deleting the disks
+        # first, so skip that if deletion of the storage account was requested.
+        if options[:storage_account]
+          delete_and_wait(sas, storage_acct_obj.name, storage_acct_obj.resource_group, options)
+        else
+          key = storage_acct_keys['key1'] || storage_acct_keys['key2']
+
+          storage_acct_obj.blobs(storage_acct_path, key).each do |blob|
+            extension = File.extname(blob.name)
+            next unless ['.vhd', '.status'].include?(extension)
+            msg = "Deleting blob #{blob.container}/#{blob.name}" if options[:verbose]
+
+            if blob.name == storage_acct_disk
+              wait_for_blob(blob)
+              Azure::Armrest::Configuration.log.info(msg) if options[:verbose]
+              storage_acct_obj.delete_blob(blob.container, blob.name, key)
+            end
+
+            if extension == '.status' && blob.name.start_with?(vm.name)
+              wait_for_blob(blob)
+              Azure::Armrest::Configuration.log.info(msg) if options[:verbose]
+              storage_acct_obj.delete_blob(blob.container, blob.name, key)
+            end
+          end
+        end
+      end
+
+      # Wait for blob to reach unlocked status.
+      def wait_for_blob(blob, max_time = 120)
+        cur_time = 0
+
+        while blob.properties.lease_status.downcase != 'unlocked'
+          cur_time += 10
+          break if cur_time > max_time
```

Comment:
Yeah, my philosophy is that if it takes more than 2 minutes for the status to update, then just let it fail because it probably means something has gone wrong.

I could just do the "skip & log" thing if you think that's more appropriate.